File: go/src/RumbleRacing/core/asset/asset.go
package asset

import "rumble-reader/chunk/shoc"

type Asset interface {
	GetType() string
	RawData() []byte
}

type GenericAsset struct {
	header  shoc.SHDR
	tag     string
	rawData []byte
}

func (g *GenericAsset) GetType() string {
	return g.tag
}

func (g *GenericAsset) RawData() []byte {
	return g.rawData
}

func (g *GenericAsset) Header() shoc.SHDR {
	return g.header
}

func ParseGenericAsset(buf []byte, tag string, header shoc.SHDR) (*GenericAsset, error) {
	resource := GenericAsset{
		header:  header,
		rawData: buf,
		tag:     tag,
	}

	return &resource, nil
}

File: src/RumbleRacing/asset/asset.ts
import ArrayBufferSlice from "../../ArrayBufferSlice";
import { SHDR } from "../chunk/shoc/shdr";

export interface Asset {
  getType(): string;
  rawData(): ArrayBufferSlice;
}

export interface GenericAsset extends Asset {
  kind: "GenericAsset";
  tag: string;
  header: SHDR;
}

export function parseGenericAsset(
  buf: ArrayBufferSlice,
  tag: string,
  header: SHDR,
): GenericAsset {
  return {
    kind: "GenericAsset",
    tag,
    header,
    getType: () => tag,
    rawData: () => buf,
  };
}
